File: src/CPCREDO.Domain/Loans/CreditPool.cs
using CPCREDO.Domain.Common;

namespace CPCREDO.Domain.Loans;

public class CreditPool
{
    public Guid Id { get; set; } = Guid.NewGuid();
    public Guid TenantId { get; set; }
    public string CurrencyCode { get; set; } = Currencies.Htg;
    public decimal FundedTotal { get; set; }
    public decimal DisbursedTotal { get; set; }
    public DateTime UpdatedAtUtc { get; set; }

    public ICollection<CreditPoolMovement> Movements { get; set; } = new List<CreditPoolMovement>();
}

public class CreditPoolMovement
{
    public Guid Id { get; set; } = Guid.NewGuid();
    public Guid TenantId { get; set; }
    public Guid PoolId { get; set; }
    public string Kind { get; set; } = "Funding";
    public string SourceKind { get; set; } = "Vault";
    public Guid? BankAccountId { get; set; }
    public Guid? LoanId { get; set; }
    public decimal Amount { get; set; }
    public string CurrencyCode { get; set; } = Currencies.Htg;
    public string? Note { get; set; }
    public Guid? JournalEntryId { get; set; }
    public Guid CreatedByUserId { get; set; }
    public DateTime CreatedAtUtc { get; set; }

    public CreditPool? Pool { get; set; }
}

File: src/CPECDO.Web/src/api/creditPool.ts
import { getToken, parseError, apiFetch } from "./client";

function headers(idem?: string): HeadersInit {
  const token = getToken();
  return {
    "Content-Type": "application/json",
    ...(token ? { Authorization: `Bearer ${token}` } : {}),
    ...(idem ? { "Idempotency-Key": idem } : {})
  };
}

export type CreditPoolMovement = {
  id: string;
  kind: string;
  sourceKind: string;
  amount: number;
  currencyCode: string;
  note: string | null;
  loanId: string | null;
  createdAtUtc: string;
};

export type CreditPool = {
  id: string;
  currencyCode: string;
  fundedTotal: number;
  disbursedTotal: number;
  reservedApprovals: number;
  availableToLend: number;
  movements: CreditPoolMovement[];
};

export async function fetchCreditPool(currency = "HTG"): Promise<CreditPool> {
  const response = await apiFetch(`/api/v1/credit-pool?currency=${currency}`, { headers: headers() });
  if (!response.ok) throw new Error(await parseError(response));
  return (await response.json()) as CreditPool;
}

export async function fundCreditPool(body: {
  sourceKind: string;
  bankAccountId?: string;
  amount: number;
  currencyCode: string;
  note?: string;
}): Promise<CreditPool> {
  const response = await apiFetch("/api/v1/credit-pool/funding", {
    method: "POST",
    headers: headers(crypto.randomUUID()),
    body: JSON.stringify(body)
  });
  if (!response.ok) throw new Error(await parseError(response));
  return (await response.json()) as CreditPool;
}
